### PR TITLE
Add item content tab rework

### PR DESCRIPTION
**Pull request by GabrielVidal1**
I think I addressed most of the changes requested in the previous pull request.
Just a few things i'm not confident about:
* I couldn't figure out how to make the content tab selected by default (in the item-details component)
* I still struggle a little bit to understand all the permissions and what they entail.

### CODE REVIEW

File: src/app/modules/item/pages/item-details/item-details.component.html [hunk 1]
```diff
@@ -8,7 +8,18 @@
...
+        [routerLink]="'./'"
+        routerLinkActive #rla1="routerLinkActive"
+        [routerLinkActiveOptions]="{exact: true}"
+        [active]="rla1.isActive">
+        Content
+      </a>
+    </nav>
+    <div class="bg-white">
+      <router-outlet></router-outlet>
+    </div>
```
Comment (smadbe) [inline src/app/modules/item/pages/item-details/item-details.component.html:22]: You haven't solved "It does not work correctly as the tab is not selected by default."
Have you check what this "exact:true" is for?

File: src/app/modules/item/services/item-datasource.service.ts [hunk 2]
```diff
@@ -22,11 +22,13 @@ export class ItemDataSource implements OnDestroy {
...
   private state = new BehaviorSubject<Ready<ItemData>|Fetching|FetchError>(fetchingState());
   state$ = this.state.asObservable();
-  item$ = this.state.pipe( // only fetched items, to be use in template as it cannot properly infer types
+  itemData$ = this.state.pipe( // only fetched items, to be use in template as it cannot properly infer types
     filter<Ready<ItemData>|Fetching|FetchError, Ready<ItemData>>(isReady),
-    map(s => s.data.item)
+    map(s => s.data)
   )
 
+  item$ = this.itemData$.pipe(map(s => s.item));
```
Comment (smadbe) [inline src/app/modules/item/services/item-datasource.service.ts:30]: No linebreak between state and itemData but one between itemdata and item?
Add semicolon to be consistent with the other lines.

**Review by smadbe — CHANGES_REQUESTED** [review]
Ok, better. Still several changes to be done.

>     * I couldn't figure out how to make the content tab selected by default (in the item-details component)
See question in code. Otherwise, I'll check myself.

>     * I still struggle a little bit to understand all the permissions and what they entail.
Just ask. 
All permissions on items are explained on https://france-ioi.github.io/algorea-devdoc/design/access-rights/items/
The user needs can_view >= 'content' to see children. So you can filter on "can_view !== 'info'" for displaying the children component.

File: src/app/modules/item/services/item-datasource.service.ts (hunk 2 above)
Comment (GabrielVidal1) [reply src/app/modules/item/services/item-datasource.service.ts:29]: fixed

File: src/app/modules/item/components/chapter-children/chapter-children.component.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      this.getItemChildrenService
+        .get(this.itemData.item.id, this.itemData.attemptId)
+        .subscribe(
+          children => {
+            this.children = children;
+            this.state = 'ready';
+        },
+          _err => {
+            this.state = 'error';
+        });
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/item/components/chapter-children/chapter-children.component.ts]: fixed

File: src/app/modules/item/components/chapter-children/chapter-children.component.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  private reloadData() {
+    if (this.itemData.attemptId) {
+      this.state = 'loading';
+      this.getItemChildrenService
+        .get(this.itemData.item.id, this.itemData.attemptId)
+        .subscribe(
+          children => {
+            this.children = children;
+            this.state = 'ready';
+        },
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/item/components/chapter-children/chapter-children.component.ts]: fixed

File: src/app/modules/item/components/chapter-children/chapter-children.component.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+
+  ngOnChanges(_changes: SimpleChanges): void {
+    this.reloadData();
+  }
+
+  private reloadData() {
+    if (this.itemData.attemptId) {
+      this.state = 'loading';
+      this.getItemChildrenService
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/item/components/chapter-children/chapter-children.component.ts]: fixed

File: src/app/modules/item/pages/item-details/item-details.component.html (hunk 1 above)
Comment (GabrielVidal1) [reply src/app/modules/item/pages/item-details/item-details.component.html]: I just removed the routerLinkActiveOptions

File: src/app/modules/item/components/chapter-children/chapter-children.component.html [hunk 6]
```diff
@@ -0,0 +1,17 @@
+<ng-container *ngIf="itemData.item.string.description !== undefined">
+  <alg-section icon="fa fa-book" label="Description" *ngIf="itemData.item.type === 'Chapter'">
+    <p *ngIf="itemData.item.string.description === null ; else noDesc">{{ itemData.item.string.description }} </p>
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/item/components/chapter-children/chapter-children.component.html]: fixed

File: src/app/modules/item/pages/item-content/item-content.component.html [hunk 7]
```diff
@@ -0,0 +1,6 @@
+<ng-container *ngIf="itemData$ | async as itemData">
+  <ng-container *ngIf="itemData.attemptId !== null">
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/item/pages/item-content/item-content.component.html]: I kept them separated for the future components that will be beside the chapter children components.

File: src/app/modules/item/components/chapter-children/chapter-children.component.ts [hunk 8]
```diff
@@ -7,14 +8,15 @@ import { ItemData } from '../../services/item-datasource.service';
...
-export class ChapterChildrenComponent implements OnChanges {
+export class ChapterChildrenComponent implements OnChanges, OnDestroy {
   @Input() itemData: ItemData;
 
   state: 'loading' | 'ready' | 'error' = 'loading';
   children: ItemChild[] = [];
 
   constructor(private getItemChildrenService: GetItemChildrenService) {}
 
+  private subscription: Subscription;
```
Comment (smadbe) [inline src/app/modules/item/components/chapter-children/chapter-children.component.ts:19]: as it does hold any value after definition and constructor, make is optional. (and handle the optional in ngOnDestroy)
```suggestion
  private subscription?: Subscription;
```

File: src/app/modules/item/components/chapter-children/chapter-children.component.ts [hunk 9]
```diff
@@ -23,7 +25,7 @@ export class ChapterChildrenComponent implements OnChanges {
   private reloadData() {
     if (this.itemData.attemptId) {
       this.state = 'loading';
-      this.getItemChildrenService
```
Comment (smadbe) [inline src/app/modules/item/components/chapter-children/chapter-children.component.ts]: The goal of the subscription is to cancel it when reloading a new item... so you need first to cancel the subscription if there is one (to cancel the request).

**Review by smadbe — CHANGES_REQUESTED** [review]
👍 
Only minor changes remaining.

File: src/app/modules/item/components/chapter-children/chapter-children.component.ts (hunk 8 above)
Comment (GabrielVidal1) [reply src/app/modules/item/components/chapter-children/chapter-children.component.ts]: fixed

File: src/app/modules/item/components/chapter-children/chapter-children.component.ts (hunk 9 above)
Comment (GabrielVidal1) [reply src/app/modules/item/components/chapter-children/chapter-children.component.ts]: fixed

File: src/app/modules/item/components/chapter-children/chapter-children.component.ts [hunk 10]
```diff
@@ -25,6 +25,8 @@ export class ChapterChildrenComponent implements OnChanges, OnDestroy {
   private reloadData() {
     if (this.itemData.attemptId) {
       this.state = 'loading';
+      if (this.subscription)
+        this.subscription.unsubscribe();
```
Comment (smadbe) [inline src/app/modules/item/components/chapter-children/chapter-children.component.ts:29]: When not using brackets, inline it. Anyway, in TS, we should use optional chaining.
```suggestion
      this.subscription?.unsubscribe();
```

File: src/app/modules/item/components/chapter-children/chapter-children.component.ts [hunk 11]
```diff
@@ -42,6 +44,7 @@ export class ChapterChildrenComponent implements OnChanges, OnDestroy {
   }
 
   ngOnDestroy() {
-    this.subscription.unsubscribe();
+    if (this.subscription)
+      this.subscription.unsubscribe();
```
Comment (smadbe) [inline src/app/modules/item/components/chapter-children/chapter-children.component.ts:48]:
```suggestion
    this.subscription?.unsubscribe();
```

**Review by smadbe — CHANGES_REQUESTED** [review]
You should use optional chaining when that's possible. (I probably forget it quite often as well).

**Review by smadbe — APPROVED** [review]

**Pull request merged**